chain-test-utils/src/time/mod.rs

```diff
@@ -0,0 +1,136 @@
+use chain_time::{Epoch, Slot, TimeEra};
+use smoke::{Generator, R};
+
+// Generate an `Epoch` from a generator function
+fn _generate_epoch<GenF>(mut gen: GenF) -> Epoch
+where
+    GenF: FnMut() -> u32,
+{
+    Epoch(gen())
+}
+
+// Generate an `Slot` from a generator function
+fn _generate_slot<GenF>(mut gen: GenF) -> Slot
+where
+    GenF: FnMut() -> u64,
+{
+    gen().into()
+}
+
+// Generate an `Epoch` given a `smoke::R` (random generator)
```

Comment:
Are we now transitioning to `smoke` from `quickcheck`, which also has a framework for arbitrary, but reproducible stuff? Does it add value to use both?

I wish I was able to see what `smoke` can do from the documentation, but there's scarcely any.

Reply:
Well, the thing with `quickcheck::Arbitrary` is that since it is a `trait` it should be implemented in the same crate as the target type. So, the idea here is to have a "centralize" crate with the utilities to generate our inner types. That way that crate can be simply used in the `tests` without relying on features. We can still use `quickcheck::Arbitrary` for the self crate tests since it is useful, but rely on `chain-test-utils` for whatever is not in that crate.

About smoke, it just has some trait for generating types and abstracts some of the most basic random numbers generators.

Reply:
> So, the idea here is to have a "centralize" crate with the utilities to generate our inner types.

That might prove a bit more difficult to maintain in separation from the crates defining the types.

> That way that crate can be simply used in the tests without relying on features.

What was the original problem we are trying to solve here? The "property-test-api" feature added the testing trait impls and possibly other test utilities for testing uses, and it was to be used in approximately this way:

```toml
# For non-test binaries.
# If the whole target is meant for testing, add "property-test-api" here as well
[dependencies]
chain-core = { path = "chain-deps/chain-core" }

# For cfg(test), integration tests under tests/, etc.
[dev-dependencies]
chain-core = { path = "chain-deps/chain-core", features = "property-test-api" }
```

Reply:
@mzabaluev the whole point of a generator based approach, is that it allow to define those generator elsewhere, since they are effectively just function/closures. A generator is just an "infinite" iterator of random value where the random engine is deterministically generated from a given seed, hence assuring reproducibility from a given `Seed`.

Reply:
I think I'm not explaining it properly, sorry.
With the `property-test-api` we are exporting the `quickcheck` dep. We are using this feature in the jormungandr tests (can't point out where exactly now). Which apparently is a misuse. 

I see your point though. We could keep the feature for testing just the `chain-deps`. (But we may have the temptation in the future to use it just because it may be easier 🍔 )

Reply:
So there are two things:
- we want a more flexible API than `arbitrary` to construct testing data;
- there is a concern about test facilities embedded into the library crates and conditionally enabled, as being cumbersome to use, or easy to get wrong and pull in unnecessary dependencies.

Would the latter be a matter of convention that has been followed only loosely, which may be what is causing problems?

Another benefit of testing APIs built into crates is that you don't have to break encapsulation to provide the testing facilities: a public generator defined in a conditionally compiled `testing` submodule can freely use the type's internals to generate values, while we don't want to provide equivalent freedom to construct values in public API.

Reply:
> Another benefit of testing APIs built into crates is that you don't have to break encapsulation to provide the testing facilities: a public generator defined in a conditionally compiled testing submodule can freely use the type's internals to generate values, while we don't want to provide equivalent freedom to construct values in public API.

Yes, this is something that will happen for sure.

Well, we can keep both though. And careful use them where they are needed.